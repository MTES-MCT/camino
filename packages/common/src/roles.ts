import { AdministrationId } from './administrations'
import { isNotNullNorUndefined } from './typescript-tools'
export const ROLES = [
  'super',
  'admin',
  'editeur',
  'lecteur',
  'entreprise',
  'bureau d’études',
  'defaut'
] as const
export type Role = typeof ROLES[number]
type UserEntreprise = { role: 'entreprise'; administrationId: undefined }
type UserBureaudEtudes = {
  role: 'bureau d’études'
  administrationId: undefined
}
type UserAdmin = { role: 'admin'; administrationId: AdministrationId }
type UserLecteur = { role: 'lecteur'; administrationId: AdministrationId }
type UserEditeur = { role: 'editeur'; administrationId: AdministrationId }
type UserSuper = { role: 'super'; administrationId: undefined }
type UserDefaut = { role: 'defaut'; administrationId: undefined }

type User =
  | { role: Role; administrationId: undefined | null | AdministrationId }
  | undefined
  | null

export const isSuper = (user: User): user is UserSuper =>
  userPermissionCheck(user, 'super')

export const isAdministration = (
  user: User
): user is UserLecteur | UserAdmin | UserEditeur =>
  isAdministrationAdmin(user) ||
  isAdministrationEditeur(user) ||
  isAdministrationLecteur(user)
export const isAdministrationAdmin = (user: User): user is UserAdmin =>
  userPermissionCheck(user, 'admin') &&
  isNotNullNorUndefined(user?.administrationId)
export const isAdministrationEditeur = (user: User): user is UserEditeur =>
  userPermissionCheck(user, 'editeur') &&
  isNotNullNorUndefined(user?.administrationId)
export const isAdministrationLecteur = (user: User): user is UserLecteur =>
  userPermissionCheck(user, 'lecteur') &&
  isNotNullNorUndefined(user?.administrationId)
export const isEntreprise = (user: User): user is UserEntreprise =>
  userPermissionCheck(user, 'entreprise')
export const isBureauDEtudes = (user: User): user is UserBureaudEtudes =>
  userPermissionCheck(user, 'bureau d’études')
export const isDefault = (user: User): user is UserDefaut | undefined =>
  !user || userPermissionCheck(user, 'defaut')

export const isRole = (role: Role | string | undefined | null): role is Role =>
  ROLES.includes(role)

function userPermissionCheck(user: User, role: Role) {
  return user?.role === role
}
