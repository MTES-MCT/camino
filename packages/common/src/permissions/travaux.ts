import { isAdministration, isSuper, User } from '../roles.js'

export const canReadTravaux = (user: User) => isSuper(user) || isAdministration(user)
