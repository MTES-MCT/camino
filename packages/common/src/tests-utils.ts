import { AdministrationRole, AdminUserNotNull, EntrepriseUserNotNull, UserDefaut, UserNotNull, UserSuper } from './roles.js'
import { AdministrationId } from './static/administrations'

export type TestUser = Pick<UserSuper, 'role'> | Pick<UserDefaut, 'role'> | Pick<AdminUserNotNull, 'role' | 'administrationId'> | Pick<EntrepriseUserNotNull, 'role' | 'entreprises'>
export const testBlankUser: Omit<UserNotNull, 'role'> = {
  id: 'id',
  email: 'email@gmail.com',
  nom: 'nom',
  prenom: 'prenom',
}
export const getTestUser = (param: { role: 'super' | 'defaut' | 'entreprise' | 'bureau d’études' } | { role: AdministrationRole; administrationId: AdministrationId }): UserNotNull => {
  switch (param.role) {
    case 'super':
    case 'defaut':
      return { ...testBlankUser, role: param.role }
    case 'admin':
    case 'lecteur':
    case 'editeur':
      return { ...testBlankUser, ...param }
    case 'entreprise':
    case 'bureau d’études':
      return { ...testBlankUser, ...param, entreprises: [] }
  }
}
