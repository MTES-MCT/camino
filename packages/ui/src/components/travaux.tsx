import { FunctionalComponent } from 'vue'
import { Page } from './demarches/page'
import { CaminoFiltre } from 'camino-common/src/filters'

export const filtres = [
  'titresIds',
  'domainesIds',
  'typesIds',
  'statutsIds',
  'entreprisesIds',
  'substancesIds',
  'references',
  'travauxTypesIds',
  'demarchesStatutsIds',
  'etapesInclues',
  'etapesExclues',
] as const satisfies readonly CaminoFiltre[]

export const Travaux: FunctionalComponent = () => {
  return <Page travaux={true} filtres={filtres} />
}
// Demandé par le router car utilisé dans un import asynchrone /shrug
Travaux.displayName = 'Travaux'
