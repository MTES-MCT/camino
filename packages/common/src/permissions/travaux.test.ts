import { test, expect } from 'vitest'
import { Role, UserNotNull } from '../roles'
import { testBlankUser } from '../tests-utils'
import { canReadTravaux } from './travaux'

const users: Record<Role, UserNotNull> = {
  super: { ...testBlankUser, role: 'super' },
  admin: { ...testBlankUser, role: 'admin', administrationId: 'aut-97300-01' },
  editeur: { ...testBlankUser, role: 'editeur', administrationId: 'aut-97300-01' },
  lecteur: { ...testBlankUser, role: 'lecteur', administrationId: 'aut-97300-01' },
  entreprise: { ...testBlankUser, role: 'entreprise', entreprises: [] },
  'bureau d’études': { ...testBlankUser, role: 'bureau d’études', entreprises: [] },
  defaut: { ...testBlankUser, role: 'defaut' },
}

test('canReadTravaux', () => {
  expect(Object.values(users).map(user => ({ role: user?.role, canReadTravaux: canReadTravaux(user) }))).toMatchSnapshot()
})
