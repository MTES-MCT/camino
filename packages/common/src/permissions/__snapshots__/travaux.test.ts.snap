// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`canReadTravaux 1`] = `
[
  {
    "canReadTravaux": true,
    "role": "super",
  },
  {
    "canReadTravaux": true,
    "role": "admin",
  },
  {
    "canReadTravaux": true,
    "role": "editeur",
  },
  {
    "canReadTravaux": true,
    "role": "lecteur",
  },
  {
    "canReadTravaux": false,
    "role": "entreprise",
  },
  {
    "canReadTravaux": false,
    "role": "bureau d’études",
  },
  {
    "canReadTravaux": false,
    "role": "defaut",
  },
]
`;
